crates/bevy_render/src/pipeline_keys/composite.rs

```diff
@@ -0,0 +1,115 @@
+use super::*;
+use bevy_ecs::schedule::NodeConfigs;
+pub use bevy_render_macros::PipelineKey;
+use bevy_utils::{intern::Interned, HashMap};
+use std::any::TypeId;
+
+#[allow(unused_imports)]
+use bevy_utils::all_tuples;
+
+#[allow(unused_macros)]
+macro_rules! impl_composite_key_tuples {
```

Comment:
I think you might have disregarded `bitbybit` a bit too quickly, as this looks very similar to the code it generates.

Reply:
The big difference is that this supports dynamic keys, where components are registered at runtime.